base_layer/core/src/transactions/transaction_components/encrypted_data.rs

```diff
@@ -84,25 +84,28 @@ impl EncryptedData {
         value: MicroTari,
         mask: &PrivateKey,
     ) -> Result<EncryptedData, EncryptedDataError> {
-        let mut openings = value.as_u64().to_le_bytes().to_vec();
-        openings.append(&mut mask.to_vec());
-        let aead_payload = Payload {
-            msg: openings.as_slice(),
-            aad: Self::TAG,
-        };
+        // Encode the value and mask
+        let mut bytes = Zeroizing::new([0u8; SIZE_VALUE + SIZE_MASK]);
+        bytes[..SIZE_VALUE].clone_from_slice(value.as_u64().to_le_bytes().as_ref());
+        bytes[SIZE_VALUE..].clone_from_slice(mask.as_bytes());
 
         // Produce a secure random nonce
-        let mut nonce = [0u8; size_of::<XNonce>()];
-        OsRng.fill_bytes(&mut nonce);
-        let nonce_ga = XNonce::from_slice(&nonce);
+        let nonce = XChaCha20Poly1305::generate_nonce(&mut OsRng);
 
+        // Set up the AEAD
         let aead_key = kdf_aead(encryption_key, commitment);
         let cipher = XChaCha20Poly1305::new(GenericArray::from_slice(aead_key.reveal()));
-        let mut ciphertext = cipher.encrypt(nonce_ga, aead_payload)?;
-        let mut ciphertext_integral_nonce = nonce.to_vec();
-        ciphertext_integral_nonce.append(&mut ciphertext);
 
-        EncryptedData::from_bytes(ciphertext_integral_nonce.as_slice())
+        // Encrypt in place
+        let tag = cipher.encrypt_in_place_detached(&nonce, ENCRYPTED_DATA_AAD, bytes.as_mut_slice())?;
+
+        // Put everything together: nonce, ciphertext, tag
+        let mut data = [0u8; SIZE_TOTAL];
+        data[..SIZE_NONCE].clone_from_slice(&nonce);
+        data[SIZE_NONCE..SIZE_NONCE + SIZE_VALUE + SIZE_MASK].clone_from_slice(bytes.as_slice());
+        data[SIZE_NONCE + SIZE_VALUE + SIZE_MASK..].clone_from_slice(&tag);
+
+        EncryptedData::from_bytes(data.as_slice())
```

Comment:
Why the additional byte copy?

```suggestion
        Ok(Self { data })
```

Reply:
Unnecessary holdover from the previous design. Will update!